conn/conn.go

```diff
@@ -0,0 +1,131 @@
+package conn
+
+import (
+	"net"
+	"time"
+)
+
+type (
+	// RLConnection is a wrapper for the net.Conn object that allows
+	// the user to limit the connection throughput.
+	RLConnection struct {
+		conn             net.Conn  // underlying net.Conn
+		packetSize       uint64    // size of single packet in bytes
+		packetsPerSecond uint64    // packets that are sent per second
+		writtenBytes     uint64    // number of bytes that have been written to the latest packet
+		lastWrite        time.Time // time when last packet was started
+	}
+)
+
+// Dial creates a RateLimitedDialer.
+func Dial(dialer *net.Dialer, network string, address string) (*RLConnection, error) {
```

Comment:
Probably better to have a `NewRLConn(conn net.Conn) *RLConn` function. That way, you can add rate limiting to connections that already exist, and you can rate-limit anything that satisfies `net.Conn` (which makes mocking easier).

Reply:
Might also be a good idea for the constructor to take `packetSize` and `packetsPerSecond` args; otherwise you have to make an extra call to `SetLimit` everywhere.